### ISSUE TITLE
Allow parameterization of Layouts

### ISSUE THREAD

**Issue by amjames** [pytorch/pytorch#80549]
### 🚀 The feature, motivation and pitch

Some layouts can be arbitrarily parameterized. Block sparse layouts, for example require block sizes to completely specify the layout. 

If we were to implement a way of parameterizing such layouts, then passing the layout in any context would be fully specified without the need for additional arguments. 

### Alternatives

- All interfaces accepting a layout would strictly support a subset of layouts.
   - This would make the adoption of any "parameterized" layout more difficult.
- All interfaces accepting a `layout` would have to accept any additional parameters required to perform any possible conversions.
   - These interfaces would need to be updated when new layouts are introduced.
   - Specification of a parameter with a `layout` option that it does not pertain to would need to be detected and handled (Warning/Error), or ignored (potentially problematic)

### Additional context

A key example (see #80542). 

If the `Tensor.to` method were to support layout conversions a `blocksize` parameter would need to be accepted, but ignored for most layouts. If the `blocksize` was instead a parameter attached to the layout itself that interface (and all others accepting a `layout`) could safely be ignorant of the additional details for those layouts. The parameter need only be accessed at the point where the real format conversion function gets involved. 

```python
x = torch.Tensor(...)
y = x.to(layout=torch.sparse_bsr(blocksize=(...)))
z = x.to(layout=torch.sparse_coo)
```
vs

```python
x = torch.Tensor(...)
y = x.to(layout=torch.sparse_bsr, blocksize=(...))
z = x.to(layout=torch.sparse_coo, blocksize=(...))
```

The handling of the last case in above would be tedious, and get worse as more formats introduce additional specialty parameters.

**Comment by pearu** [comment]
^ @cpuhrsch

**Comment by pearu** [comment]
I agree that it is not clear how to parameterize Layout's using the current implementation of the layout feature but see the idea below.

That said, it would be natural to specify dense dimensions in a sparse/sparse_csr/... layouts as well. Currently, it is impossible to construct an empty sparse hybrid tensor using `torch.empty(..., layout=torch.sparse...)` approach. The same note applies to other tensor factory functions.

A rough idea (Python users only): Currently, the `layout` argument must be `torch.layout` instance when specified. We could generalize this to allow the `layout` to be an arbitrary object that must implement `layout` property/attribute returning `torch.layout` instance but the object could also carry other information (`blocksize`, `densesize`, etc) that will be internally used when constructing the tensor object with the given layout. For instance:
```python
class Layout:
    def __init__(self, layout, blocksize=(), densesize=()):
        self.layout = layout
        self.blocksize = blocksize
        self.densesize = densesize

t = torch.empty((2, 3), layout=Layout(torch.sparse_csr, densesize=(4, 5)))
```
where `t` would be a sparse hybrid CSR tensor with two dense dimensions.
That is, we would not need to change the current `layout` enum-based implementation but the idea is about overloading the `layout` argument with a possibility to pass other parameters to the factory functions without the need to introduce new layout-specific arguments to the existing factory functions.

**Comment by pearu** [comment]
The Index idea from https://github.com/pytorch/pytorch/issues/76638#issuecomment-1118892355 that could cover the `blocksize` part of the issue.
